groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -765,787 +1431,10708 @@ void Optional_DataImp<TYPE>::emplace(
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator==(const bsl::optional<LHS_TYPE>& lhs,
+                const std::optional<RHS_TYPE>& rhs);
+    // If bool(x) != bool(y), false; otherwise if bool(x) == false, true;
+    // otherwise *x == *y. Note that this function will fail to compile if
+    // 'LHS_TYPE' and 'RHS_TYPE' are not compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator!=(const bsl::optional<LHS_TYPE>& lhs,
+                const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator!=(const std::optional<LHS_TYPE>& lhs,
+                const bsl::optional<RHS_TYPE>& rhs);
+    // If bool(x) != bool(y), true; otherwise, if bool(x) == false, false;
+    // otherwise *x != *y.  Note that this function will fail to compile if
+    // 'LHS_TYPE' and 'RHS_TYPE' are not compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator<(const bsl::optional<LHS_TYPE>& lhs,
+               const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator<(const std::optional<LHS_TYPE>& lhs,
+               const bsl::optional<RHS_TYPE>& rhs);
+    // If !y, false; otherwise, if !x, true; otherwise *x < *y.  Note that this
+    // function will fail to compile if 'LHS_TYPE' and 'RHS_TYPE' are not
+    // compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator>(const bsl::optional<LHS_TYPE>& lhs,
+               const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator>(const std::optional<LHS_TYPE>& lhs,
+               const bsl::optional<RHS_TYPE>& rhs);
+    // If !x, false; otherwise, if !y, true; otherwise *x > *y. note that this
+    // function will fail to compile if 'LHS_TYPE' and 'RHS_TYPE' are not
+    // compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator<=(const bsl::optional<LHS_TYPE>& lhs,
+                const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator<=(const std::optional<LHS_TYPE>& lhs,
+                const bsl::optional<RHS_TYPE>& rhs);
+    // If !x, true; otherwise, if !y, false; otherwise *x <= *y. Note that this
+    // function will fail to compile if 'LHS_TYPE' and 'RHS_TYPE' are not
+    // compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator>=(const bsl::optional<LHS_TYPE>& lhs,
+                const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator>=(const std::optional<LHS_TYPE>& lhs,
+                const bsl::optional<RHS_TYPE>& rhs);
+    // If !y, true; otherwise, if !x, false; otherwise *x >= *y. Note that this
+    // function will fail to compile if 'LHS_TYPE' and 'RHS_TYPE' are not
+    // compatible.
+
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+
+template <class TYPE>
+BSLS_KEYWORD_CONSTEXPR bsl::optional<typename bsl::decay<TYPE>::type>
+alloc_optional(typename bsl::optional<
```

Comment:
there is an e-mail thread on this. The alloc vs allocate is an unintentional misnomer I somehow managed to introduce when writing the paper. It never got fixed because Pablo raised a naming issue which we never really resolved. I'll leave this comment open until this is resolved.

Reply:
decided to change these to be allocator extended overloads of make_optional. For now, they are in the same place where they used to be because I think that will make the difference easier to spot. However, I think it would be better if allocator extended make_optional is after the "plan" make_optional. Leaving this comment open to remind me to re-arrange the order

Reply:
It's an interesting idea.  It is, however, different from the other places in BDE where such constructs are found.  For example, `shared_ptr` has `make_shared` for using default allocator and `allocate_shared(Allocator *, ARGS)` for non-default one.  We also have similar duality in `makeManaged` and `allocateManaged` for our `ManagedPtr`.  

I think it is different for optional though.  With smart pointers, there are 2 things going on - (1) allocation of the object footprint on the heap using the allocator and (2) allocator use of the object itself.  With `optional` it is different, because there's no (1).  This actually brings me to a question - don't we have the same problem that we had with `value_or` when guaranteed copy elision is not available?  I mean, one can do `make_optional<string>(allocator_arg, alloc, "cat")` and if copy elision doesn't kick in, they will get an optional having default allocator, right?  

If we forget the issue above for a moment, I think I'd still prefer to have `allocate_optional`.  `make_optional` helps us get rid of use of `in_place`.  Just as a matter of usability, I think `allocate_optional` should help us get rid of both `allocator_arg` and `in_place`.

If these considerations make sense to you, perhaps this is a discussion that we might want to include Mike, Pablo, and other interested parties (Alisdair?).  If that's the case, please send an email.

Reply:
Yup, we do have the same problem as with 'value_or'. I'm glad you brought it up because t crossed my mind a few times, but never when I was at my desk. :/  I think I would want to put ifguards around it, just like we did with 'value_or'. 

I prefer the' allocate_optional' over 'make_optional overload', but my arguments were not strong enough to convince Pablo. Pablo's biggest concern with 'allocate_optional' is that there is no(1) and therefore  the name is wrong ('allocate_optional' does not allocate anything). I'm sympathetic to that. 
My biggest concern is that 'make_optional' now does two different things. One set of overloads says "give me an optional with a value created from these arguments" (i.e., all these arguments are the same), and the other set says "give me an optional with value created from these arguments *and* use this other argument  for the optional itself"(i.e. treat this one argument differently to others). 
This is because the allocator argument isn't just forwarded blindly with all the other arguments. It needs to be recognised so it can be tacked on before the 'in_place' tag. I'm not sure I'm verbalising it well, and it's really just a niggle. Ville likens it to vector<bool>.

I think I don't feel particularly strongly about this. I would want a symmetry with other types, but I am  sufficiently convinced that 'allocate_optional' is the wrong name. 'make_optional_with_allocator' is a possibility, but that doesn't address the symmetry. I'll probably send out an email anyway. The fact that Ville, you and I all prefer the 'allocate_optional' says something. If nothing else, it says that there are likely to be others of the same opinion, and we will need to address this when talking to the committee.

Reply:
going back to the return value optimization and 'make_optional' with an allocator - it would appear that all the compilers I have tested so far are able to optimize the return value in allocator extended 'make_optional' case (there is no issue in plain 'make_optional', you get default allocator no matter what you do). My guess is that this is because 'make_optional' is a very simple case. I would recommend we leave it in as is. If we find that there is a compiler on which tests for 'make_optional' fail, we can revisit this decision. WDYT ?